capstone/scripts/convert_s3.py

```diff
@@ -89,123 +89,135 @@ def export_cases_to_s3(redacted: bool, reporter_id: str) -> None:
+    vars["fake_request"] = namedtuple("Request", ["query_params", "accepted_renderer"])(
+        query_params={"body_format": "text"},
+        accepted_renderer=None,
+    )
+
+    # create a metadata contents string to append case metadata content
+    metadata_contents = b""
+
+    # store the serialized case data in tempfile
+    for item in cases_search.scan():
+        with tempfile.NamedTemporaryFile() as file:
+            # pass case in to add additional data to the CaseDocument
+            case = cases_by_id[item["_source"]["id"]]
 
-    # select corresponding case search obj
-    case_search = CaseDocument.raw_search().filter("term", id=case.id)
-    # Store the serialized data
-    with tempfile.NamedTemporaryFile() as file:
-        for item in case_search.scan():
             serializer = vars["serializer"](
-                item["_source"], context={"request": vars["fake_request"]}
+                item["_source"],
+                context={
+                    "request": vars["fake_request"],
+                    "first_page_order": case.first_page_order,
+                    "last_page_order": case.last_page_order,
+                },
             )
-            file.write(json.dumps(serializer.data).encode("utf-8") + b"\n")
+
+            # add data to metadata_contents string without 'casebody'
+            metadata_data = serializer.data
+            metadata_data.pop("casebody", None)
+            metadata_contents += json.dumps(metadata_data).encode("utf-8") + b"\n"
+
+            # compose each case file with a hash
+            case_contents = json.dumps(serializer.data).encode("utf-8") + b"\n"
+            hash_object = hashlib.sha256(case_contents)
+            case_contents_hash = base64.b64encode(hash_object.digest()).decode()
+
+            # calculate case file name
+            if prev_case_first_page == case.first_page:
+                case_file_name_index += 1
+            else:
+                case_file_name_index = 1
+            case_file_name = (
+                f"{case.first_page.zfill(4)}-{str(case_file_name_index).zfill(2)}.json"
+            )
+
+            # set so we can use to determine multiple cases on single page
+            prev_case_first_page = case.first_page
+
+            # ignore cases that have already been uploaded
+            if (
+                case_contents_hash in s3_contents_hash
+                and s3_contents_hash[case_contents_hash] == case_file_name
+            ):
+                s3_contents_hash.pop(case_contents_hash, None)
+                print(f"Skipped {cases_key}{case_file_name}")
+                continue
+            else:
+                file.write(case_contents)
+                file.flush()
+                hash_and_upload(
+                    file,
+                    dest_bucket,
+                    f"{cases_key}{case_file_name}",
+                    "application/jsonl",
+                )
+
```

Comment:
Here we have to delete anything left in `s3_contents_hash` from s3; otherwise making a change to the export format and rerunning will result in double files.

Reply:
I'm confused here. Here's the situation: 
Right now we query S3 for files and hashes and hold in a dict.
If a case's hash is the same as an S3 file hash and the names are the same, pop it from the dict and don't upload.
Otherwise, write a new S3 file from the case contents.

If I'm reading you correctly, what I think you're pointing out is:
When there's already a case of that name on S3, but the file hashes are different, we should delete that file from S3 and write a new one.

So that would look to me like:
if hashes are the same and file names are the same, pop from dict and don't upload (continue)
else if file names are the same but hashes are not the same, delete from S3, and upload new file
else upload new file.

Is that the correct interpretation?

Reply:
Michael helped get to a different understanding. We now think you're saying: pop if the name and file contents are the same, if they're not upload and then pop, and at the end if you have anything else left, delete those.

Reply:
How about something like this?

```
s3_existing_object_hashes = {'key': 'hash', ...}  # map keys to hashes

for item in cases_search.scan():
    # do upload if file changed
    dest_key = f"{cases_key}{case_file_name}"
    s3_contents_hash = s3_existing_object_hashes.pop(dest_key)
    if s3_contents_hash is None or s3_contents_hash != case_contents_hash:
        ... # do upload

for extra_s3_object in s3_existing_object_hashes:
    ... # delete extra_s3_object
```

We push the contents for any file that doesn't have the same contents already existing at that key. And then if any keys aren't encountered at all, we know they're an extra file and need to be deleted.